pyroma/projectdata.py

```diff
@@ -3,9 +3,69 @@
 import sys
 import logging
 import tokenize
-
+import build
+import email
+import pathlib
+import tempfile
 from copy import copy
 from distutils import core
+from setuptools import config
+
+METADATA_MAP = {
+    "summary": "description",
+    "classifier": "classifiers",
+    "project_url": "project_urls",
+    "home_page": "url",
+}
+
+
+def get_build_data(path):
+    with tempfile.TemporaryDirectory() as tempdir:
+        metadata_dir = build.ProjectBuilder(str(path)).prepare("wheel", tempdir)
+        with open(pathlib.Path(metadata_dir) / "METADATA", "rb") as metadata_file:
+            metadata = email.message_from_binary_file(metadata_file)
+
+    # This method tends to add two newlines to the description, we clean that up here:
+    long_description = metadata.get_payload().strip() + "\n"
+    data = {'long_description': long_description}
+    for key, value in metadata.items():
+        if value.strip() == "UNKNOWN":
+            continue
+        key = key.lower().replace('-', '_')
+        if key in METADATA_MAP:
+            key = METADATA_MAP[key]
+        if key in data:
+            if isinstance(data[key], list):
+                data[key].append(value)
+            else:
+                data[key] = [data[key], value]
+        else:
+            data[key] = value
+    return data
+
+
+def get_setupcfg_data(path):
+    data = config.read_configuration("setup.cfg")
+    metadata = data["metadata"]
+    return metadata
+
+
+def get_data(path):
+    try:
+        return get_build_data(path)
+    except build.BuildException:
+        # It couldn't build the package, this happens if you only have a setup.cfg and
+        # no setup.py, which I'm testing for, but maybe it's not actually supported?
+        return get_setupcfg_data(path)
+    except Exception:
+        logging.exception("Build failed, try falling back to monkey-patching setup.py")
+        metadata = get_setuppy_data(path)
+        logging.exception(
+            "Fallback worked! This likely means your package is using very old packaging techniques, "
+            "and Pyroma will soon regard that as a complete failure and rate you as not even cheese.\n"
+            "Please modernize your packaging!"
```

Comment:
I'd be very curious if this ever actually triggers in practice. The previous should catch the very small fraction of projects that have dumped `setup.py` but forgot to actually set the correct `build-system` section in the `pyproject.toml`; otherwise, for those have the former but are missing the latter, `build` (like all standards-conforming tools) assumes `setuptools.build_meta:__legacy__` (i.e. direct `setup.py` invocation), which _should_ provide close to identical behavior when requesting the metadata as the deprecated monkeypatching here (so either both will fail, or neither will). And if PEP 517 builds are used and `build-system` is properly configured, metadata retrieval should basically boil down to reading the `setup.cfg` or `pyproject.toml`, or at most executing the initial stages of the `setup.py` in a similar fashion as here.

That said, its not impossible there are cases that could trigger this; I'd be very eager to learn what they are, if so.

Reply:
It triggers with ooooold Distutils packages, for example. So I think this will only stay for a short while. But I've gotten a lot of flack for not being backwards compatible on other packages, so...

Reply:
I'd be really curious to see one...would be like a trip back in time, like reading the `pep2html.py` script in the Python PEPs repo, that we just now finally got rid of in favor of our new Sphinx based build system described in PEP 676. Some parts of it had been untouched for _20 years_...it was like a time capsule to the early days of Python. Of course, that also made it an absolute nightmare to work with, heh...

Reply:
I think what is happening here is that the setup.py fails, but my patched setuptools call succeeds. So most likely this is a broken setup py. So I'm adding a warning for that for this version and will just fail in the future.